includes/GadgetResourceLoaderModule.php

```diff
@@ -88,6 +94,262 @@ public function getTargets() {
+		}
+
+		$services = MediaWikiServices::getInstance();
+		$revisionLookup = $services->getRevisionLookup();
+
+		$latestRev = $revisionLookup->getRevisionByTitle( $title );
+
+		$latestMainSlotContent = $this->getMainSlotContent( $latestRev );
+
+		if ( !$latestMainSlotContent ) {
+			return null;
+		}
+
+		$mainSlotRoleHandler = $services->getSlotRoleRegistry()->getRoleHandler( SlotRecord::MAIN );
+
+		// CSS pages are not subject to ContentReview
+		// NOTE: Determine the content model solely based on the title and ignore the actual revision content model
+		// so that the review process cannot be bypassed by changing the content model of a JS page to CSS
+		if ( $mainSlotRoleHandler->getDefaultModel( $title ) === CONTENT_MODEL_CSS ) {
+			return $latestMainSlotContent;
+		}
+
+		$liveRevisionId = $services->getService( ContentReviewStorage::class )->getLiveRevisionForPage(
+			$services->getMainConfig()->get( 'CityId' ),
+			$title->getArticleID()
+		);
+
+		$liveRevision = $revisionLookup->getRevisionById( $liveRevisionId );
+
+		return $this->getMainSlotContent( $liveRevision );
+	}
+
+	/**
+	 * Fandom change - IW-4592 - Gadgets ContentReview integration
+	 * Get the main slot content for the given revision, returning null if the revision doesn't exist
+	 * or the main slot content is a redirect.
+	 *
+	 * @param RevisionRecord|null $revision
+	 * @return Content|null
+	 */
+	private static function getMainSlotContent( ?RevisionRecord $revision ): ?Content {
+		if ( !$revision ) {
+			return null;
+		}
+
+		$mainSlotContent = $revision->getContent( SlotRecord::MAIN, RevisionRecord::RAW );
+
+		if ( $mainSlotContent->isRedirect() ) {
+			return null;
+		}
+
+		return $mainSlotContent;
+	}
+
+	/**
+	 * Fandom change - IW-4592 - Gadgets ContentReview integration
+	 * Fetch metadata for the pages that compose this module, for freshness checks.
+	 * @param ResourceLoaderContext $context
+	 * @return array|mixed
+	 * @throws ConfigException
+	 */
+	protected function getTitleInfo( ResourceLoaderContext $context ) {
+		$pageNames = array_keys( $this->getPages( $context ) );
+		sort( $pageNames );
+		$batchKey = implode( '|', $pageNames );
+
+		if ( !isset( $this->titleInfo[$batchKey] ) ) {
+			$this->titleInfo[$batchKey] = $this->loadTitleInfo( $context, $pageNames, __METHOD__ );
+		}
+
+		return $this->titleInfo[$batchKey];
+	}
+
+	/**
+	 * Fandom change - IW-4592 - Gadgets ContentReview integration
+	 * Load page revision metadata for freshness checks.
+	 * If ContentReview is active and not in test mode, use the metadata of the latest reviewed revisions for JS pages.
+	 *
+	 * @param ResourceLoaderContext $context
+	 * @param array $pages
+	 * @param string $fname
+	 * @return array - map of page names to corresponding metadata
+	 * @throws ConfigException
+	 * @see ResourceLoaderWikiModule::fetchTitleInfo()
+	 *
+	 */
+	protected function loadTitleInfo( ResourceLoaderContext $context, array $pages, $fname = __METHOD__ ) {
```

Comment:
I suggest to split this function into smaller well described functions...it's pretty hard to understand what is going on...+ there are not tests here...